libc/string_mem.c

```diff
@@ -0,0 +1,391 @@
+/*
+ * Phoenix-RTOS
+ *
+ *    libc-tests
+ *    HEADER:
+ *    - string.h
+ *    TESTED:
+ *    - memset(),
+ *    - memmove()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <ulimit.h>
+#include <stdint.h>
+#include <unity_fixture.h>
+
+#define BUF_LEN         32
+#define BIG_SIZE        1024
+#define TEST_STR        "THIS IS TEST MEM123"
+#define TEST_STR_SIZE   sizeof(TEST_STR)
+#define OFFSET          40
+#define OFFSET_BIG      1500
+#define MEM_SIZE        100
+#define MEM_SIZE_BIG    4000
+#define OVERLAP_LEN     (TEST_STR_SIZE / 2)
+#define OVERLAP_LEN_BIG (BIG_SIZE / 2)
+
+static uint8_t buf[BUF_LEN];
+static uint8_t buf_byte[256];
+static uint8_t buf_byte_exp[256];
+static uint8_t *ret;
+static uint8_t *mem;
+static uint8_t *mem_big;
+static uint8_t big_buf[1024];
+static const int values[] = { 0, 1, 16, 50, 127, 128, 200, 255,
+	768, 257, 272, 306, 383, 384, 456, 511 };
+
+#define VALUES_NUM sizeof(values) / sizeof(values[0])
+
+
+static void setup_mem_big(int offset)
+{
+	memset(mem_big, 0, MEM_SIZE_BIG);
+	uint8_t *dst = mem_big + OFFSET_BIG;
+	TEST_ASSERT_NOT_NULL(mem_big);
+	ret = memcpy(dst + offset, big_buf, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(dst + offset, ret);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem_big, OFFSET_BIG + offset);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, dst + offset + BIG_SIZE, MEM_SIZE_BIG - (OFFSET_BIG + BIG_SIZE + offset));
+}
+
+
+TEST_GROUP(string_memset);
+TEST_GROUP(string_memmove);
+TEST_GROUP(string_memmove_big);
+
+
+TEST_SETUP(string_memset)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memset)
+{
+}
+
+
+TEST(string_memset, basic)
+{
+	int i;
+	for (i = 0; i < VALUES_NUM; i++) {
+		ret = memset(buf, values[i], BUF_LEN);
+		TEST_ASSERT_EQUAL_PTR(buf, ret);
+		TEST_ASSERT_EACH_EQUAL_INT8((uint8_t)values[i], buf, BUF_LEN);
+	}
+}
+
+
+TEST(string_memset, zero)
+{
+	uint8_t val = 5;
+	ret = memset(&val, 0, 0);
+	TEST_ASSERT_EQUAL_PTR(&val, ret);
+	TEST_ASSERT_EQUAL_UINT8(5, val);
+}
+
+
+TEST(string_memset, big)
+{
+	int i, j, n = BIG_SIZE - 2 * 8;
+	for (j = 0; j < 8; j++) {
+		for (i = 0; i < 8; i++) {
+			ret = memset(big_buf + i, 10, n + j);
+			TEST_ASSERT_EQUAL_PTR(big_buf + i, ret);
+			TEST_ASSERT_EACH_EQUAL_UINT8(10, big_buf + i, n + j);
```

Comment:
not resetting data between the tests, not checking if more than requested data was overwritten